lib/mayaUsd/ufe/UsdTransform3dCommonAPI.cpp

```diff
@@ -42,32 +43,31 @@ class UsdTranslateUndoableCmd : public Ufe::TranslateUndoableCommand
                 item->prim().GetPath().GetText());
         }
 
-        _prevT = t;
         _t = t;
     }
 
-    void undo() override { _commonAPI.SetTranslate(_prevT, _time); }
-    void redo() override { _commonAPI.SetTranslate(_t, _time); }
-
     // Executes the command by setting the translation onto the transform op.
     bool set(double x, double y, double z) override
     {
         _t = GfVec3d(x, y, z);
-        redo();
+        execute();
```

Comment:
I'm perplexed by the fact that set() calls execute(), which will call executeUndoBlock().  set() is called during the manipulation; at that point whatever attributes needed to be added to the prim to receive the in-manipulation values should have been added.  If you call execute() (and thus executeUndoBlock()) isn't it the case that you will replace the contents of the UsdUndoableItem with the new value you've set?  See here:
https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/undo/UsdUndoManager.cpp#L56
This is insufficient to do a complete undo, e.g. if you've added a new translate xform op to the prim, it should be removed.

Reply:
But we do need to capture the new value in the UsdUndoableItem so that if the user does undo then redo, the redo puts the latest value that was set, right?

I'm not sure how that would work. If we only call set, then the UsdUndoableItem will not contain the latest value? But OTOH, as you say, if we call execute, then if the attribute had been created, the newly generated UsdUndoableItem will not capture that.

I'll have to think about how to make both cases work.

Reply:
I'll take a look at the API of UsdUndoableItem to see if we can update its content without creating a new instance from scratch.

Reply:
Ah, I see that doInvert() declares an UsdUndoBlock, so that will capture the latest value once undo() gets called. I had assume the undo item was capture both the old and new value. So I'll change the code not to call execute.

Reply:
Looking at the code and running the unit-tests, I saw that the correct fix was to keep calling execute() but having execute() detects that it had already been executed and not refill the UsdUndoableItem, instead skipping it and directly calling the implementation (executeUndoBlock).

So, for short, for these interactive commands, set(0 can call execute() which will take care of capturing the changes within a UsdUndoBlock + UsdUndoableItem or not, depending on the fact that execute() has already been called or not, which was conveniently already captured in the state variable.